### DISCUSSION TITLE
Laravel Starter Kit proposal

### DISCUSSION THREAD

**Discussion by woodseowl** [CornellCustomDev/CD-LaravelStarterKit#2] [Announcements]
#### Rationale and Motivation
When we build Laravel applications in Custom Development, we deliver enterprise solutions that meet standards and expectations for quality, security, accessibility, branding, performance, timeliness, and support. We can improve our delivery of that work with consistent, high-quality development, deployment, and maintenance practices. Creating and maintaining a common starting point for our Laravel applications that incorporates our best approaches will help us routinely and more rapidly deliver solutions with these qualities. It will also establish practices that foster team learning and continuous improvement of our work.
 
#### Goals

- Reduce time to build Laravel apps
- Increase consistency in configuration, third-party packages, and architecture
- Increase code visibility, code quality, and team collaboration
- Continuously improve code and practices
- Lower barriers for support and reduce support time

 
#### Phases
- Kick-off
  - Set up a mini-project and organize the people and plans to make it happen
- Draft Requirements
  - Brainstorm feature requirements
  - Brainstorm process and quality requirements to support ongoing development
  - Review Laravel best practices for starter kits
  - Find Custom Dev projects that can be used as examples of feature requirements
  - Gather insights from CD Demo and CSS Framework process
  - Document requirements and resources for Initial Design
  - Team review and revisions of requirements 
- Initial Design
  - Propose v1.0 features
  - Propose v1.0 candidate architecture, base packages, and code to port from existing project
  - Propose planning, development, and support process
  - Team review and revisions of initial design
- Initial Build
  - Find a new Custom Dev project that can be used for creating a starter kit
  - Create a v0.1 starter kit (MVP)
  - Generate new project from starter kit
  - Iterate improvements back from project to starter kit when possible
  - Review and document process and quality improvements with team
- Full Release
  - Continue generating new projects from starter kit
  - Iterate improvements back from projects to starter until v1.0 features are met
  - Review and document process and quality improvements with team
- Support Hand Off
  - Draft a roadmap
  - Create support plan based on iterative process
  - Team review and revisions

#### Estimates

- Kick-off: 4h (team meeting)
- Draft Requirements: 20h + team review
- Initial Design: TBD based on requirements + team review
- Initial Build: TBD based on design + project-funded time + team review
- Full Release: TBD + project-funded time + team review
- Support Hand Off: TBD + team review
- Maintenance: TBD hours / period

**Comment by inaydich** [comment]
I remember I did a lot of css components clean-up with IrinaZ help in GSC (https://github.com/CU-CommunityApps/CD-gradcmte). I think it is a good project to use as an example.

**Comment by richmarisa** [comment]
GPS took components from EntryExit and Daily Check to build forms, datatables from GSC.  I was looking at the Alpine UI component library... I bought myself a license but that may not work for the group.

**Comment by woodseowl** [comment]
Per Nov. 2 meeting (#11), I updated the Draft Requirements milestone to start with brainstorming tasks and added a documentation task at the end.

**Reply by woodseowl** [reply]
Alternatively (from Slack): "Maybe we should take a radically iterative approach to the starter kit and our v0.1 is just Laravel, then we start proposing modifications? So really what we would be working on is our process and standards for iterative / collaborative development."

**Comment by woodseowl** [comment]
Per Nov. 15 meeting (#18), working iteratively is a good approach. This means that the original proposal is too heavy on big planning up front, so I'll work on revising the proposal.
